File: Client-Python/command_processor.py
from typing import List

class CommandProcessor:
    SPACE = " "
    BGN_TUPLE = "<"
    END_TUPLE = ">"
    PUT = "put"
    READ = "read"
    TAKE = "take"
    EXIT = "exit"
    GET_TUPLE_SPACES_STATE = "getTupleSpacesState"
    SLEEP = "sleep"

    def __init__(self, client_service):
        self.client_service = client_service

    def parse_input(self):
        exit_flag = False
        while not exit_flag:
            try:
                line = input("> ").strip()
                split = line.split()
                if not split:
                    continue

                command = split[0]

                if command == self.PUT:
                    print(self.put(split))
                elif command == self.READ:
                    print(self.read(split))
                elif command == self.TAKE:
                    print(self.take(split))
                elif command == self.GET_TUPLE_SPACES_STATE:
                    print(self.get_tuple_spaces_state())
                elif command == self.SLEEP:
                    self.sleep(split)
                elif command == self.EXIT:
                    exit_flag = True
                else:
                    self.print_usage()
            except EOFError:
                break

    def put(self, split: List[str]):
        if not self.input_is_valid(split):
            return self.print_usage()
        tuple_data = split[1]
        delays = self.extract_delays(split)
        return self.client_service.put(tuple_data, delays)

    def read(self, split: List[str]):
        if not self.input_is_valid(split):
            return self.print_usage()
        tuple_data = split[1]
        delays = self.extract_delays(split)
        return self.client_service.read(tuple_data, delays)

    def take(self, split: List[str]):
        if not self.input_is_valid(split):
            return self.print_usage()
        tuple_data = split[1]
        delays = self.extract_delays(split)
        return self.client_service.take(tuple_data, delays)

    def get_tuple_spaces_state(self):
        return self.client_service.get_tuple_spaces_state()

    def sleep(self, split: List[str]):
        if len(split) != 2:
            return self.print_usage()
        try:
            import time
            time.sleep(int(split[1]))
        except ValueError:
            return self.print_usage()

    def extract_delays(self, split: List[str]) -> List[int]:
        try:
            return [int(split[i]) if i < len(split) else 0 for i in range(2, 5)]
        except ValueError:
            return [0, 0, 0]

    def print_usage(self):
        print("Usage:\n"
              "- put <tuple> [delay0 delay1 delay2]\n"
              "- read <tuple> [delay0 delay1 delay2]\n"
              "- take <tuple> [delay0 delay1 delay2]\n"
              "- getTupleSpacesState\n"
              "- sleep <seconds>\n"
              "- exit\n")

    def input_is_valid(self, split: List[str]) -> bool:
        if len(split) < 2 or len(split) > 5:
            return False

        # Verifica se o tuplo começa com < e termina com >
        if not split[1].startswith(self.BGN_TUPLE) or not split[1].endswith(self.END_TUPLE):
            return False

        # Verifica se todos os delays (se existirem) são inteiros >= 0
        for delay_arg in split[2:]:
            try:
                value = int(delay_arg)
                if value < 0:
                    return False
            except ValueError:
                return False

        return True

File: Client-Java/src/main/java/pt/ulisboa/tecnico/tuplespaces/client/CommandProcessor.java
package pt.ulisboa.tecnico.tuplespaces.client;

import java.util.List;
import java.util.Scanner;

import pt.ulisboa.tecnico.tuplespaces.client.grpc.ClientService;

public class CommandProcessor {

    private static final String SPACE = " ";
    private static final String BGN_TUPLE = "<";
    private static final String END_TUPLE = ">";
    private static final String PUT = "put";
    private static final String READ = "read";
    private static final String TAKE = "take";
    private static final String SLEEP = "sleep";
    private static final String EXIT = "exit";
    private static final String GET_TUPLE_SPACES_STATE = "getTupleSpacesState";

    private final ClientService clientService;

    public CommandProcessor(ClientService clientService) {
        this.clientService = clientService;
    }

    void parseInput() {

        Scanner scanner = new Scanner(System.in);
        boolean exit = false;

        while (!exit) {
            System.out.print("> ");
            String line = scanner.nextLine().trim();
            String[] split = line.split(SPACE);
            switch (split[0]) {
                case PUT:
                    this.put(split);
                    break;

                case READ:
                    this.read(split);
                    break;

                case TAKE:
                    this.take(split);
                    break;

                case GET_TUPLE_SPACES_STATE:
                    this.getTupleSpacesState();
                    break;

                case SLEEP:
                    this.sleep(split);
                    break;

                case EXIT:
                    exit = true;
                    break;

                default:
                    this.printUsage();
                    break;
            }
        }
        scanner.close();
    }

    private void put(String[] split) {
        if (!this.inputIsValid(split)) {
            System.out.println("Invalid input");
            this.printUsage();
            return;
        }

        String tuple = split[1];
        int[] delays = extractDelays(split);

        String result = clientService.put(tuple, delays);
        System.out.println(result + "\n");
        }

    private void read(String[] split) {
        if (!this.inputIsValid(split)) {
            System.out.println("Invalid input");
            this.printUsage();
            return;
        }

        String tuple = split[1];
        int[] delays = extractDelays(split);

        String result = clientService.read(tuple, delays);
        System.out.println(result + "\n");
    }

    private void take(String[] split) {
        // check if input is valid
        if (!this.inputIsValid(split)) {
            System.out.println("Invalid input");
            this.printUsage();
            return;
        }

        // get the tuple
        String tuple = split[1];

        // Call ClientService.take
        int[] delays = extractDelays(split);
        String result = clientService.take(tuple, delays);

        System.out.println(result + "\n");
    }

    private void getTupleSpacesState() {
        List<String> tupleSpacesState = clientService.getTupleSpacesState();
        System.out.println(tupleSpacesState + "\n");

    }

    private void sleep(String[] split) {
        if (split.length != 2) {
            this.printUsage();
            return;
        }
        Integer time;

        // checks if input String can be parsed as an Integer
        try {
            time = Integer.parseInt(split[1]);
        } catch (NumberFormatException e) {
            this.printUsage();
            return;
        }

        try {
            Thread.sleep(time * 1000);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Extrai os valores de delay do comando inserido pelo utilizador.
     */
    private int[] extractDelays(String[] split) {
        int[] delays = {0, 0, 0}; // Default: 0 para todas as réplicas
        int delayCount = Math.min(3, split.length - 2);

        for (int i = 0; i < delayCount; i++) {
            try {
                delays[i] = Integer.parseInt(split[2 + i]);
            } catch (NumberFormatException e) {
                delays[i] = 0;
            }
        }
        return delays;
    }

    private void printUsage() {
        System.out.println("Usage:\n" +
                "- put <element[,more_elements]>\n" +
                "- read <element[,more_elements]>\n" +
                "- take <element[,more_elements]>\n" +
                "- getTupleSpacesState <server>\n" +
                "- sleep <integer>\n" +
                "- exit\n");
    }

    private boolean inputIsValid(String[] input) {
        // Tem que ter pelo menos o comando e o tuplo
        if (input.length < 2 || input.length > 5) {
            return false;
        }

        String tuple = input[1];
        if (!tuple.startsWith(BGN_TUPLE) || !tuple.endsWith(END_TUPLE)) {
            return false;
        }

        // Verificar se os argumentos adicionais (delays) são inteiros >= 0
        for (int i = 2; i < input.length; i++) {
            try {
                int delay = Integer.parseInt(input[i]);
                if (delay < 0) {
                    return false;
                }
            } catch (NumberFormatException e) {
                return false;
            }
        }

        return true;
    }

}
